### PR TITLE
libflux: clean up module API

### PR DESCRIPTION
**Pull request by garlick**
This PR culls some module management functions from the API that were not all that helpful, as proposed in #1854.  It also scraps a problematic unit test and replaces it with one that ought to work better.  Finally, functions are made more robust to bad arguments.

This module stuff requires more work to eliminate the log to stderr in `flux_modname()` and to add dynamically registered service names.  I started to work on both of those and both turned into a much bigger change than I was comfortable with taking on right now.

This should not be merged until a flux-sched PR goes in, to wean it from functions that are being removed here.

### CODE REVIEW

**Comment by garlick** [comment]
On listing dynamically registered service names - actually it wasn't that big of a deal so I added it.
`flux module list` now shows services, minus the implicit service that matches the module name.
```
$ flux module list -r all
Module               Size    Digest  Idle  S   Nodeset Service
job                  1182184 41FE11D    0  S     [0-3] 
resource-hwloc       1127680 03725D7    0  S     [0-3] 
userdb               1108216 6FD7788    0  S         0 
cron                 1193656 D2E75FE    0  S         0 
barrier              1106088 18A39F2    0  S     [0-3] 
content-sqlite       1115896 297E6EE    0  S         0 content-backing
kvs-watch            1338368 10B4AD1    0  S     [0-3] 
aggregator           1125232 77C3D23    0  S     [0-3] 
connector-local      1155136 CCDB76A    0  R     [0-3] 
kvs                  1613360 B1DC8F4    0  S     [0-3] 
```

In addition to the sched update, RFC 5 will need to add a services array to the lsmod response.  It can be empty if the base module does not offer dynamic service registration.

File: src/common/libflux/test/module.c [hunk 1]
```diff
@@ -0,0 +1,132 @@
+#include <sys/param.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <string.h>
+#include <errno.h>
+
+#include "src/common/libflux/module.h"
+#include "src/common/libtap/tap.h"
+
+
```
[the remark this answers is not in the document]
Comment (garlick) [reply src/common/libflux/test/module.c]: Great point, I'll add that.

**Comment by garlick** [comment]
> Should there be a new test for listing the module services? AFAICT there isn't.

Yup, will add that.

> should this be two PRs? One for the cleanup and one for the new service listing feature?

Well since both changes involved editing the same code at point of use, it seemed good to combine and get it over with, especially since sched needs to be syned.

I thought of a way to address the stderr logging in `flux_modname()`, `flux_modfind()` too, so I'll go ahead and push that and return to address your comments.  Thanks!

**Comment by garlick** [comment]
Just forced a push after adding a test that services are listed in `flux module list`, adding a comment in the unit test about the FAKE1 and FAKE2 defines, and adding the "dlerror callback" idea to replace stderr logging in `flux_modname()`, which seems to have worked out OK.

**Comment by grondo** [comment]
> `flux module list` now shows services, minus the implicit service that matches the module name.

Hm, output from `flux module list -r all` seems to attribute services registered only on one rank to all ranks, which might be confusing. E.g. in this example I've registered `myservice` on rank 0 using `connector-local`:

```
grondo@peep:/tmp/f.git$ flux module list -r 0 | grep local
connector-local      1154992 4E3BF6D    0  R         0 myservice
grondo@peep:/tmp/f.git$ flux module list -r 1 | grep local
connector-local      1154992 4E3BF6D    5  S         1 
grondo@peep:/tmp/f.git$ flux module list -r all | grep local
connector-local      1154992 4E3BF6D    0  R     [0-3] myservice
```

I'm not sure how big of a deal this is, and from your comment on `lsmod_services_string` this seems to be intentional. However, it might be less misleading if `flux module list -r all` split lines with different services, e.g. in this case:

```
connector-local      1154992 4E3BF6D    0  R         0 myservice
connector-local      1154992 4E3BF6D    0  R     [1-3]
```

**Comment by garlick** [comment]
Thanks, I was being lazy there I guess, but what you propose certainly makes more sense.  Just pushed a change that should produce the results you described.  Will need to add some tests.

**Comment by garlick** [comment]
Some incremental devel sqaushed.  Ready I think?

**Comment by garlick** [comment]
Thanks!

**Comment by garlick** [comment]
It could probably be skipped since the dispatch code already checks that,
although this does follow the typical pattern if a request handler so IMHO
prob doesn't hurt to retain that pattern.

**Pull request merged**